Panic on assignment to entry in nil map at `streamsMap.AddStream`

On new session [`handleConn`](https://github.com/masih/webtransport-go/blob/72455c5fd07be29cacc64e9b1b99435c881dc060/session.go#L114) eventually [closes the session](https://github.com/masih/webtransport-go/blob/72455c5fd07be29cacc64e9b1b99435c881dc060/session.go#L130) executed in a goroutine. As part of this closure, `streamsMap` [is set to `nil`](https://github.com/masih/webtransport-go/blob/f892c8f3d7806ca5bae004ac9c2f533876883ba9/streams_map.go#L41).

I wonder if it is possible for session manager to return a session that is already closed, which could then result in a panic below.

```
panic: assignment to entry in nil map
goroutine 2004647120 [running]:
github.com/quic-go/webtransport-go.(*streamsMap).AddStream(0xc194fdbde8, 0xc15f2954a0?, 0xc025dc8b70)
	/go/pkg/mod/github.com/quic-go/webtransport-go@v0.5.1/streams_map.go:24 +0x66
github.com/quic-go/webtransport-go.(*Session).addStream(0xc194fdbd10, {0x7f2b66fb6cf0?, 0xc15f2954a0}, 0x60?)
	/go/pkg/mod/github.com/quic-go/webtransport-go@v0.5.1/session.go:173 +0x152
github.com/quic-go/webtransport-go.(*Session).OpenStreamSync(0xc194fdbd10, {0x136cb60?, 0xc087ad9920?})
	/go/pkg/mod/github.com/quic-go/webtransport-go@v0.5.1/session.go:324 +0x21e
github.com/libp2p/go-libp2p/p2p/transport/webtransport.(*conn).OpenStream(0x4b23f034feb1?, {0x136cb60?, 0xc087ad9920?})
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.25.1/p2p/transport/webtransport/conn.go:48 +0x2b
github.com/libp2p/go-libp2p/p2p/net/swarm.(*Conn).openAndAddStream(0xc0d09e8ab0, {0x136cb60?, 0xc087ad9920?}, {0x1376500, 0x1d05f40})
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.25.1/p2p/net/swarm/swarm_conn.go:220 +0x42
github.com/libp2p/go-libp2p/p2p/net/swarm.(*Conn).NewStream(0xc0d09e8ab0, {0x136cb60, 0xc087ad9920})
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.25.1/p2p/net/swarm/swarm_conn.go:211 +0x175
github.com/libp2p/go-libp2p/p2p/net/swarm.(*Swarm).NewStream(0xc000726000, {0x136cb60, 0xc087ad9920}, {0xc2252abad0, 0x26})
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.25.1/p2p/net/swarm/swarm.go:401 +0x1f0
github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).NewStream(0xc00027e780, {0x136cb60, 0xc087ad9920}, {0xc2252abad0, 0x26}, {0xc000868070?, 0x1, 0x1})
	/go/pkg/mod/github.com/libp2p/go-libp2p@v0.25.1/p2p/host/basic/basic_host.go:626 +0x16f
github.com/libp2p/go-libp2p-kad-dht/internal/net.(*peerMessageSender).prep(0xc23c3a3ae0, {0x136cb60?, 0xc087ad9920?})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/internal/net/message_manager.go:218 +0x6a
github.com/libp2p/go-libp2p-kad-dht/internal/net.(*peerMessageSender).prepOrInvalidate(0xc23c3a3ae0, {0x136cb60, 0xc087ad9920})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/internal/net/message_manager.go:200 +0x91
github.com/libp2p/go-libp2p-kad-dht/internal/net.(*messageSenderImpl).messageSenderForPeer(0xc000d88e00, {0x136cb60, 0xc087ad9920}, {0xc2252abad0, 0x26})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/internal/net/message_manager.go:150 +0x233
github.com/libp2p/go-libp2p-kad-dht/internal/net.(*messageSenderImpl).SendRequest(0xc000d88e00, {0x136cab8, 0xc16b6a9b40}, {0xc2252abad0, 0x26}, 0xe50620?)
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/internal/net/message_manager.go:78 +0xd8
github.com/libp2p/go-libp2p-kad-dht/pb.(*ProtocolMessenger).GetClosestPeers(0xc000868130, {0x136cab8, 0xc16b6a9b40}, {0xc2252abad0, 0x26}, {0xc0633607b0?, 0x0?})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/pb/protocol_messenger.go:108 +0xd4
github.com/libp2p/go-libp2p-kad-dht.(*IpfsDHT).FindPeer.func1({0x136cab8, 0xc16b6a9b40}, {0xc2252abad0, 0x26})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/routing.go:604 +0xfb
github.com/libp2p/go-libp2p-kad-dht.(*query).queryPeer(0xc1920f7cb0, {0x136cab8, 0xc16b6a9b40}, 0xc02ce7d120?, {0xc2252abad0, 0x26})
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/query.go:408 +0x26e
created by github.com/libp2p/go-libp2p-kad-dht.(*query).spawnQuery
	/go/pkg/mod/github.com/libp2p/go-libp2p-kad-dht@v0.21.0/query.go:325 +0x44a
```

--- a/session.go
+++ b/session.go
@@ -297,7 +297,7 @@ rand:
 	return id
 }
 
-func (s *Session) OpenStreamSync(ctx context.Context) (str Stream, err error) {
+func (s *Session) OpenStreamSync(ctx context.Context) (Stream, error) {
 	s.closeMx.Lock()
 	if s.closeErr != nil {
 		s.closeMx.Unlock()
@@ -307,19 +307,22 @@ func (s *Session) OpenStreamSync(ctx context.Context) (str Stream, err error) {
 	id := s.addStreamCtxCancel(cancel)
 	s.closeMx.Unlock()
 
-	defer func() {
-		s.closeMx.Lock()
-		defer s.closeMx.Unlock()
-		delete(s.streamCtxs, id)
-		if err != nil && s.closeErr != nil {
-			err = s.closeErr
-		}
-	}()
-
-	var qstr quic.Stream
-	qstr, err = s.qconn.OpenStreamSync(ctx)
+	qstr, err := s.qconn.OpenStreamSync(ctx)
 	if err != nil {
+		if s.closeErr != nil {
+			return nil, s.closeErr
+		}
 		return nil, err
+	}
+
+	s.closeMx.Lock()
+	defer s.closeMx.Unlock()
+	delete(s.streamCtxs, id)
+	// Some time might have passed. Check if the session is still alive
+	if s.closeErr != nil {
+		qstr.CancelWrite(sessionCloseErrorCode)
+		qstr.CancelRead(sessionCloseErrorCode)
+		return nil, s.closeErr
 	}
 	return s.addStream(qstr, true), nil
 }
@@ -348,19 +351,21 @@ func (s *Session) OpenUniStreamSync(ctx context.Context) (str SendStream, err er
 	id := s.addStreamCtxCancel(cancel)
 	s.closeMx.Unlock()
 
-	defer func() {
-		s.closeMx.Lock()
-		defer s.closeMx.Unlock()
-		delete(s.streamCtxs, id)
-		if err != nil && s.closeErr != nil {
-			err = s.closeErr
-		}
-	}()
-
-	var qstr quic.SendStream
-	qstr, err = s.qconn.OpenUniStreamSync(ctx)
+	qstr, err := s.qconn.OpenUniStreamSync(ctx)
 	if err != nil {
+		if s.closeErr != nil {
+			return nil, s.closeErr
+		}
 		return nil, err
+	}
+
+	s.closeMx.Lock()
+	defer s.closeMx.Unlock()
+	delete(s.streamCtxs, id)
+	// Some time might have passed. Check if the session is still alive
+	if s.closeErr != nil {
+		qstr.CancelWrite(sessionCloseErrorCode)
+		return nil, s.closeErr
 	}
 	return s.addSendStream(qstr), nil
 }

--- a/session_test.go
+++ b/session_test.go
@@ -117,3 +117,62 @@ func TestAddStreamAfterSessionClose(t *testing.T) {
 	ustr.EXPECT().CancelRead(sessionCloseErrorCode)
 	sess.addIncomingUniStream(ustr)
 }
+
+func TestOpenStreamAfterSessionClose(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockSess := NewMockStreamCreator(ctrl)
+	mockSess.EXPECT().Context().Return(context.WithValue(context.Background(), quic.ConnectionTracingKey, uint64(1337)))
+	wait := make(chan struct{})
+	mockSess.EXPECT().OpenStreamSync(gomock.Any()).DoAndReturn(func(context.Context) (quic.Stream, error) {
+		str := NewMockStream(ctrl)
+		str.EXPECT().CancelRead(sessionCloseErrorCode)
+		str.EXPECT().CancelWrite(sessionCloseErrorCode)
+		<-wait
+		return str, nil
+	})
+
+	sess := newSession(42, mockSess, newMockRequestStream(ctrl))
+
+	errChan := make(chan error, 1)
+	go func() {
+		_, err := sess.OpenStreamSync(context.Background())
+		errChan <- err
+	}()
+
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, sess.CloseWithError(0, "session closed"))
+
+	close(wait)
+	require.EqualError(t, <-errChan, "session closed")
+}
+
+func TestOpenUniStreamAfterSessionClose(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockSess := NewMockStreamCreator(ctrl)
+	mockSess.EXPECT().Context().Return(context.WithValue(context.Background(), quic.ConnectionTracingKey, uint64(1337)))
+	wait := make(chan struct{})
+	mockSess.EXPECT().OpenUniStreamSync(gomock.Any()).DoAndReturn(func(context.Context) (quic.SendStream, error) {
+		str := NewMockStream(ctrl)
+		str.EXPECT().CancelWrite(sessionCloseErrorCode)
+		<-wait
+		return str, nil
+	})
+
+	sess := newSession(42, mockSess, newMockRequestStream(ctrl))
+
+	errChan := make(chan error, 1)
+	go func() {
+		_, err := sess.OpenUniStreamSync(context.Background())
+		errChan <- err
+	}()
+
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, sess.CloseWithError(0, "session closed"))
+
+	close(wait)
+	require.EqualError(t, <-errChan, "session closed")
+}

--- a/session_test.go
+++ b/session_test.go
@@ -125,7 +125,9 @@ func TestOpenStreamAfterSessionClose(t *testing.T) {
 	mockSess := NewMockStreamCreator(ctrl)
 	mockSess.EXPECT().Context().Return(context.WithValue(context.Background(), quic.ConnectionTracingKey, uint64(1337)))
 	wait := make(chan struct{})
+	streamOpen := make(chan struct{})
 	mockSess.EXPECT().OpenStreamSync(gomock.Any()).DoAndReturn(func(context.Context) (quic.Stream, error) {
+		streamOpen <- struct{}{}
 		str := NewMockStream(ctrl)
 		str.EXPECT().CancelRead(sessionCloseErrorCode)
 		str.EXPECT().CancelWrite(sessionCloseErrorCode)
@@ -140,8 +142,8 @@ func TestOpenStreamAfterSessionClose(t *testing.T) {
 		_, err := sess.OpenStreamSync(context.Background())
 		errChan <- err
 	}()
+	<-streamOpen
 
-	time.Sleep(100 * time.Millisecond)
 	require.NoError(t, sess.CloseWithError(0, "session closed"))
 
 	close(wait)
@@ -155,7 +157,9 @@ func TestOpenUniStreamAfterSessionClose(t *testing.T) {
 	mockSess := NewMockStreamCreator(ctrl)
 	mockSess.EXPECT().Context().Return(context.WithValue(context.Background(), quic.ConnectionTracingKey, uint64(1337)))
 	wait := make(chan struct{})
+	streamOpen := make(chan struct{})
 	mockSess.EXPECT().OpenUniStreamSync(gomock.Any()).DoAndReturn(func(context.Context) (quic.SendStream, error) {
+		streamOpen <- struct{}{}
 		str := NewMockStream(ctrl)
 		str.EXPECT().CancelWrite(sessionCloseErrorCode)
 		<-wait
@@ -169,8 +173,8 @@ func TestOpenUniStreamAfterSessionClose(t *testing.T) {
 		_, err := sess.OpenUniStreamSync(context.Background())
 		errChan <- err
 	}()
+	<-streamOpen
 
-	time.Sleep(100 * time.Millisecond)
 	require.NoError(t, sess.CloseWithError(0, "session closed"))
 
 	close(wait)

